### ISSUE TITLE
Capturing aexpect log messages

### ISSUE THREAD

**Issue by luckyh** [avocado-framework/avocado-vt#3369]
We used to have the aexpect log messages presented in `debug.log` which help us a lot for debugging automation issues. After the bug [1] had been fixed, however, the aexpect log messages was gone from that file due to it was using the root logger. Now it is still missing although there was already a fix [2] for changing its logging namespace to its own one.

After some investigation I discovered that (at least) there were two approaches to get the aexpect log messages back:

1. using the particular Avocado option mentioned in the doc [3]. But there is an issue that the format of those logs were missing a critical field - timestamps, and also I'm not even sure if that is the expected behavior, or is it?

```
[stdlog] 2022-03-14 02:15:09,746 avocado.virttest.virt_vm DEBUG| Found/Verified IP 192.168.1.6 for VM avocado-vt-vm1 NIC 0
[stderr] DEBUG:aexpect.client:Sending command: ls
[stderr] 
[stderr] DEBUG:aexpect.client:Sending command: echo $?   
[stderr] 
[stdlog] 2022-03-14 02:15:12,310 avocado.virttest.qemu_vm DEBUG| Destroying VM avocado-vt-vm1 (PID 1623055)
```

2. applying a workaround(?) to Avocado-VT as the following.

```patch
diff --git a/avocado_vt/test.py b/avocado_vt/test.py
index d04c09fe..636ca5d9 100644
--- a/avocado_vt/test.py
+++ b/avocado_vt/test.py
@@ -21,6 +21,7 @@ import logging
 import os
 import sys
 import pipes
+from contextlib import contextmanager
 
 from avocado.core import exceptions
 from avocado.core import test
@@ -69,6 +70,34 @@ def cleanup_env(env_filename, env_version):
     env.destroy()
 
 
+@contextmanager
+def capture_extra_log_messages(test):
+    def reconfigure_logger(name, level=logging.DEBUG):
+        logger = logging.getLogger(name)
+        base_level = logger.level
+        logger.setLevel(level)
+        base_propagate = logger.propagate
+        logger.propagate = False
+        for handler in test.log.handlers:
+            if handler not in logger.handlers:
+                logger.addHandler(handler)
+        return (base_level, base_propagate)
+
+    def restore_logger(name, base_values):
+        logger = logging.getLogger(name)
+        logger.setLevel(base_values[0])
+        logger.propagate = base_values[1]
+        for handler in test.log.handlers:
+            if handler in logger.handlers:
+                logger.removeHandler(handler)
+
+    aexpect_base_values = reconfigure_logger("aexpect")
+    try:
+        yield
+    finally:
+        restore_logger("aexpect", aexpect_base_values)
+
+
 class VirtTest(test.Test, utils.TestUtils):
 
     """
@@ -144,7 +173,8 @@ class VirtTest(test.Test, utils.TestUtils):
         env_lang = os.environ.get('LANG')
         os.environ['LANG'] = 'C'
         try:
-            self._runTest()
+            with capture_extra_log_messages(self):
+                self._runTest()
             self.__status = "PASS"
         # This trick will give better reporting of virt tests being executed
         # into avocado (skips, warns and errors will display correctly)
```

```
[stdlog] 2022-03-14 06:16:12,184 avocado.virttest.virt_vm DEBUG| Found/Verified IP 192.168.1.6 for VM avocado-vt-vm1 NIC 0
[stdlog] 2022-03-14 06:16:14,159 aexpect.client DEBUG| Sending command: ls
[stdlog] 2022-03-14 06:16:14,270 aexpect.client DEBUG| Sending command: echo $?
[stdlog] 2022-03-14 06:16:14,522 avocado.virttest.qemu_vm DEBUG| Destroying VM avocado-vt-vm1 (PID 1635223)
```

For the better user experience I'd prefer the second approach, but I do concern about if it is something that would not encourage to do. @beraldoleal @richtja @clebergnu @ana please let me know your opinion, thanks!

[1] https://github.com/avocado-framework/avocado/issues/5049
[2] https://github.com/avocado-framework/aexpect/pull/89
[3] https://avocado-framework.readthedocs.io/en/95.0/guides/writer/chapters/logging.html#using-store-logging-stream

**Comment by luckyh** [comment]
@pevogam @dzhengfy @kylazhang @chunfuwen I'd like to add you here for you may also be interested in this issue, thanks!

**Comment by beraldoleal** [comment]
@luckyh. Yes, indeed. Avocado is only caring about `avocado.*` namespaces.

What about using `avocado.aexpect` as the namespace for the aexpect? Since aexpect is part of Avocado project, I don't see a problem with that.

Would that fix the issue?

**Comment by beraldoleal** [comment]
> As aexpect could be used in standalone applications (we have some usage like that) and doesn't have any explicit direct dependency on Avocado, I would recommend not mixing the namespaces any further. 

I see your point, but I don't get it as a bad thing. At the end, aexpect is part of avocado, regardless of being used as standalone. 
  
> What we do on our side to make the aexpect messages available again (but we needed the namespace [fix above](https://github.com/avocado-framework/aexpect/pull/89) first) is
> 
> ```python
> log.getLogger("aexpect").parent = log.getLogger("avocado.extlib")
> ```

IIUC, basically this is setting 'avocado.extlib.aexpect' in your side, instead of 'avocado.aexpect' by default. What is also acceptable. However...

> i.e. we bind the loggers dynamically. In particular for Avocado VT, we have to do this not just for aexpect but any other external library out there (we attach all of them to the avocado.extlib namespace above) and do this using a VT test runner customization hook from #3242. So I consider our solution an approach (3) to the two approaches above, and at least from what I can see it seems the minimal amount of changes.

The question is: Do we need to store all external library log messages? IMO an avocado/avocado-vt user should care about avocado/avocado-vt logs. As developers, we could hide implementation details and give them high level messages to improve the quality of our logs. Avoiding huge logs with external libraries messages (that we don't control). Yes, we could include external libraries errors as part of our errors, but as "advanced feature", builtin in a much more digestible Avocado log.

My concerns are:
  1. We don't control third-party libraries logs. If at some point a requirement change the messages (by increasing or reducing the number of messages) we will be affected.
  2. We are teaching our users that everything that is external it is under `avocado.extlib`. By doing that we might assume some messages will be there just because currently it is. And that might be false, if at some point in the future extlib A decided to change the log message. Users will miss that message or get a different message.
  3. By assuming third-party libraries are already logging stuff we might keep as it is and do not improve the log on our side.

One of the reasons we decided to split things it was just to avoid a huge debug.log and third-party messages that we don't control. And thanks to the Python logging system, advanced users could still get those by setting what they would like to see (and seems to me, maybe this is what @pevogam is doing). I'm just trying to thing as a non advanced user would get this.

In other words: I like the idea of giving advanced users the flexibility to configure the log as they wish. I just don't like much the idea to assume that all our users are advanced and we are going to use that by default. Because of that I would not recommend to add all third-party library logs into any 'avocado.*' namespace.

**Comment by pevogam** [comment]
> > As aexpect could be used in standalone applications (we have some usage like that) and doesn't have any explicit direct dependency on Avocado, I would recommend not mixing the namespaces any further.
> 
> I see your point, but I don't get it as a bad thing. At the end, aexpect is part of avocado, regardless of being used as standalone.

I am not sure I see your argument here. What is the "part of" supposed to mean more specifically? As I said before aexpect is a standalone application with its own packaging, unit testing, and overall usage. It is a dependency of avocado and it is under the same github umbrella but it is a different and independent tool. The method above generalized over various external and additional dependencies for Avocado VT and hard-coding a name into the logging of aexpect is going to solve just one case and not the rest.

> > What we do on our side to make the aexpect messages available again (but we needed the namespace [fix above](https://github.com/avocado-framework/aexpect/pull/89) first) is
> > ```python
> > log.getLogger("aexpect").parent = log.getLogger("avocado.extlib")
> > ```
> 
> IIUC, basically this is setting 'avocado.extlib.aexpect' in your side, instead of 'avocado.aexpect' by default. What is also acceptable. However...
> 
> > i.e. we bind the loggers dynamically. In particular for Avocado VT, we have to do this not just for aexpect but any other external library out there (we attach all of them to the avocado.extlib namespace above) and do this using a VT test runner customization hook from #3242. So I consider our solution an approach (3) to the two approaches above, and at least from what I can see it seems the minimal amount of changes.
> 
> The question is: Do we need to store all external library log messages?

Why is this even a question though? We are not trying to answer this about Avocado and the users decide about what logging messages they want to include. What I think @luckyh is trying to say is that Avocado VT lacks the aexpect logging messages and since Avocado VT is meant as an application, its maintainers and other end-users of Avocado want to decide which logging messages to include and which not.

> IMO an avocado/avocado-vt user should care about avocado/avocado-vt logs. As developers, we could hide implementation details and give them high level messages to improve the quality of our logs. Avoiding huge logs with external libraries messages (that we don't control). Yes, we could include external libraries errors as part of our errors, but as "advanced feature", builtin in a much more digestible Avocado log.

I think instead that the avocado/avocado-vt users should have the choice which messages to include and which not. If users use plugins with various additional dependencies that play crucial role in their tests (we among them as I have said multiple times before) they might actually want to add the respective logging output for their debuging.

The option I described above does exactly that: it allows the users to choose which libraries they want to include. Not allowing aexpect logging lines as lines outside of avocado is just as restrictive as demanding aexpect lines to be displayed when some of us might not want to see them for supposed high level logging output without that many details.

> My concerns are:
> 
>     1. We don't control third-party libraries logs. If at some point a requirement change the messages (by increasing or reducing the number of messages) we will be affected.

So I guess we are on more or less as similar page here and the question is just whether aexpect should be considered as such.

>     2. We are teaching our users that everything that is external it is under `avocado.extlib`. By doing that we might assume some messages will be there just because currently it is. And that might be false, if at some point in the future extlib A decided to change the log message. Users will miss that message or get a different message.

I am not sure what is critical in what you say and the "extlib" part could be changed but I found it a neat way to be more organized about logging messages from various external libraries.

>     3. By assuming third-party libraries are already logging stuff we might keep as it is and do not improve the log on our side.

I was merely suggesting a minimal pluggable way to provide a logging from external library like aexpect. I don't understand what you mean with all these interactions in mind but if Avocado VT can be considered an application where most of us build their tests (and frankly the previous logging was much easier to comprehend and created far less trouble from various disappearing messages) then I see nothing wrong about its users and maintainers making taking care of and decisions about the logging of additional dependencies that it uses (like aexpect here and possibly more for more specialized modules under the virttest directory).

> One of the reasons we decided to split things it was just to avoid a huge debug.log and third-party messages that we don't control. And thanks to the Python logging system, advanced users could still get those by setting what they would like to see (and seems to me, maybe this is what @pevogam is doing). I'm just trying to think as a non advanced user would get this.
> 
> In other words: I like the idea of giving advanced users the flexibility to configure the log as they wish. I just don't like much the idea to assume that all our users are advanced and we are going to use that by default. Because of that I would not recommend to add all third-party library logs into any 'avocado.*' namespace.

IMO the simplest logging we could have was the one we had before when no one has to constantly think which libraries are included and which are missing and Avocado was behaving more like an application capable of running tests and less like a library that should be imported. The logging just worked and no one had to think hard about various complications and ways to make things more modular. However, the change was undertaken and we had to spend a few months fixing effects in various plugins and userspace code. So now that we are finally more stable and done with this the most we can do is to protect this modularity by not adding various assumptions about stuff beyond Avocado. Avocado VT then simply doesn't have to make assumptions beyond Avocado VT usage and as it uses aexpect as a dependency (it is already encoded in VT's assumptions in order to function), Avocado VT should be free to handle the logging of the additional libraries it makes use of, hopefully in a clean minimal way like the one shown above.

**Comment by beraldoleal** [comment]
Hi @pevogam, thanks for the comments. I guess we are more or less on the same page here, it is just about the default behavior I' m talking about and also because we are talking about two things (aexpect vs others) So, let me try to answer your questions...

> > I see your point, but I don't get it as a bad thing. At the end, aexpect is part of avocado, regardless of being used as standalone.
> 
> I am not sure I see your argument here. What is the "part of" supposed to mean more specifically? As I said before aexpect is a standalone application with its own packaging, unit testing, and overall usage. It is a dependency of avocado and it is under the same github umbrella but it is a different and independent tool. The method above generalized over various external and additional dependencies for Avocado VT and hard-coding a name into the logging of aexpect is going to solve just one case and not the rest.

"Part of the same umbrella", since it is inside 'avocado-framework' repository. Would be nice from the 'propaganda PoV' to announce "avocado.aexpect". But I'm not against having as standalone 'aexpect'. I'm just saying that I don't see as a problem if it was inside avocado.* namespace. We are going on a direction to split repos, (i.e  utils repository). And if we use the same logic here we probably will remove "avocado.*" from the name space as well, what I would prefer to not remove. 
Again, I'm not against, I'm just giving my personal preference here.

> > > What we do on our side to make the aexpect messages available again (but we needed the namespace [fix above](https://github.com/avocado-framework/aexpect/pull/89) first) is
> > > ```python
> > > log.getLogger("aexpect").parent = log.getLogger("avocado.extlib")
> > > ```
> > 
> > 
> > IIUC, basically this is setting 'avocado.extlib.aexpect' in your side, instead of 'avocado.aexpect' by default. What is also acceptable. However...
> > > i.e. we bind the loggers dynamically. In particular for Avocado VT, we have to do this not just for aexpect but any other external library out there (we attach all of them to the avocado.extlib namespace above) and do this using a VT test runner customization hook from #3242. So I consider our solution an approach (3) to the two approaches above, and at least from what I can see it seems the minimal amount of changes.
> > 
> > 
> > The question is: Do we need to store all external library log messages?
> 
> Why is this even a question though? We are not trying to answer this about Avocado and the users decide about what logging messages they want to include. What I think @luckyh is trying to say is that Avocado VT lacks the aexpect logging messages and since Avocado VT is meant as an application, its maintainers and other end-users of Avocado want to decide which logging messages to include and which not, here include aexpect's logging messages.

It is a genuine question from me, @pevogam.  It is not clear to me why include 'third-party' messages in our logs by default (I'm not saying that is the case for aexpect, I'm talking about *any other external library out there*). I agree with you that users should decide what include. But IMO Avocado VT it is not only an application, it is a plugin that is part of Avocado project. Many Avocado users don't see much a difference, and depending on our policy for third-party library logs, this will affect Avocado/Avocado-VT users.
 
If aexpect's logs are crucial for avocado-vt, and since it is "part of avocado umbrella" I don't see a problem with shipping a default behavior where those logs are included (avocado.aexpect or avocado.extlib.aexpect). I got the original question from @luckyh. My issue is with *any other external library out there*.

> > IMO an avocado/avocado-vt user should care about avocado/avocado-vt logs. As developers, we could hide implementation details and give them high level messages to improve the quality of our logs. Avoiding huge logs with external libraries messages (that we don't control). Yes, we could include external libraries errors as part of our errors, but as "advanced feature", builtin in a much more digestible Avocado log.
> 
> I think instead that the avocado/avocado-vt users should have the choice which messages to include and which not. If users use plugins with various additional dependencies that play crucial role in their tests (we among them as I have said multiple times before) they might actually want to add the respective logging output for their debugging.

We are on the same page that users should choose. I'm just trying discussing the default behavior, since many users may not want to choose. The main complaint that I got many times from QEMU/Libvirt upstream community it was that debugging inside Avocado (and +Avocado-VT) is hard, and shouldn't be. I agree 100%. I'm just suggesting that we could avoid a default behavior where third-party logs are included.
 
> The option I described above does exactly that: it allows the users to choose which libraries they want to include. Not allowing aexpect logging lines as lines outside of avocado is just as restrictive as demanding aexpect lines to be displayed when some of us might not want to see them for supposed high level logging output without that many details.

I have a feeling that changing their parent could be a little "intrusive" and not needed. With the Python logging system, maybe we don't need to change their parent to 'avocado.' or 'avocado.extlib'. We just need a 'logging.conf' (or an equivalent python code) that will redirect specific libraries to our handlers (files) of interest. Again, I'm talking about 'other libraries', aexpect might be a special case.

> > My concerns are:
> > ```
> > 1. We don't control third-party libraries logs. If at some point a requirement change the messages (by increasing or reducing the number of messages) we will be affected.
> > ```
> 
> So I guess we are on more or less as similar page here and the question is just whether aexpect should be considered as such.

Indeed. Since it is on the same umbrella I guess we do have a little of control over it. That is the reason I see aexpect as an special case.

> > ```
> > 2. We are teaching our users that everything that is external it is under `avocado.extlib`. By doing that we might assume some messages will be there just because currently it is. And that might be false, if at some point in the future extlib A decided to change the log message. Users will miss that message or get a different message.
> > ```
> 
> I am not sure what is critical in what you say and the "extlib" part could be changed but I found it a neat way to be more organized about logging messages from various external libraries.

Nothing wrong with the "extlib" part. Ignore it. I will rephrase:

2. We are teaching our users that everything that is external it is under `avocado.*`. By doing that we might assume some messages will be there just because currently it is. And that might be false, if at some point in the future any external library decided to change the log message. Users will miss that message or get a different message.

> > ```
> > 3. By assuming third-party libraries are already logging stuff we might keep as it is and do not improve the log on our side.
> > ```
> 
> I was merely suggesting a minimal pluggable way to provide a logging from external library like aexpect. I don't understand what you mean with all these interactions in mind but if Avocado VT can be considered an application where most of us build their tests (and frankly the previous logging was much easier to comprehend and created far less trouble from various disappearing messages) then I see nothing wrong about its users and maintainers taking care of and making decisions about the logging of additional dependencies that it uses (like aexpect here and possibly more for more specialized modules under the virttest directory).

I guess we are on the same page, it is just a matter of the default behavior for avocado-vt/avocado users vs what users could configure, as I mentioned before. Also I don't think we should change the parent logger of an external library to do that. Avocado will care about 'avocado.*' namespaces, and users *might*, in some cases, want to choose to add 'others' namespace to their logs.

> > One of the reasons we decided to split things it was just to avoid a huge debug.log and third-party messages that we don't control. And thanks to the Python logging system, advanced users could still get those by setting what they would like to see (and seems to me, maybe this is what @pevogam is doing). I'm just trying to think as a non advanced user would get this.
> > In other words: I like the idea of giving advanced users the flexibility to configure the log as they wish. I just don't like much the idea to assume that all our users are advanced and we are going to use that by default. Because of that I would not recommend to add all third-party library logs into any 'avocado.*' namespace.
> 
> IMO the simplest logging we could have was the one we had before when no one has to constantly think which libraries are included and which are missing and Avocado was behaving more like an application capable of running tests and less like a library that should be imported. The logging just worked and no one had to think hard about various complications and ways to make things more modular. However, the change was undertaken and we had to spend a few months fixing effects in various plugins and userspace code. So now that we are finally more stable and done with this the most we can do is to protect this modularity by not adding various assumptions about stuff beyond Avocado. Avocado VT then simply doesn't have to make assumptions beyond Avocado VT usage and as it uses aexpect as a dependency (it is already encoded in VT's assumptions in order to function), Avocado VT should be free to handle the logging of the additional libraries it makes use of, hopefully in a clean minimal way like the one shown above.

I guess each plugin is free to handle the logging the way they want. And I'm just trying to discuss here alternatives, in order to make things consistent with the design of the "new logging system" adopted inside Avocado. Otherwise we are going to have two different designs in two really close projects.

**Comment by pevogam** [comment]
> > Why is this even a question though? We are not trying to answer this about Avocado and the users decide about what logging messages they want to include. What I think @luckyh is trying to say is that Avocado VT lacks the aexpect logging messages and since Avocado VT is meant as an application, its maintainers and other end-users of Avocado want to decide which logging messages to include and which not, here include aexpect's logging messages.
> 
> It is a genuine question from me, @pevogam. It is not clear to me why include 'third-party' messages in our logs by default (I'm not saying that is the case for aexpect, I'm talking about _any other external library out there_). I agree with you that users should decide what include. But IMO Avocado VT it is not only an application, it is a plugin that is part of Avocado project. Many Avocado users don't see much a difference, and depending on our policy for third-party library logs, this will affect Avocado/Avocado-VT users.
> 
> If aexpect's logs are crucial for avocado-vt, and since it is "part of avocado umbrella" I don't see a problem with shipping a default behavior where those logs are included (avocado.aexpect or avocado.extlib.aexpect). I got the original question from @luckyh. My issue is with _any other external library out there_.

I see @beraldoleal but I don't quite get where we suggest adding other external libraries here. The above is just a solution to add aexpect and its logging. If nothing else then adding aexpect in such fashion is also suggestive for downstream users about ways in which they can enable logging for their own downstream dependencies if they so desire. In short, the above is an example of a minimal modular way to add aexpect logging and possibly even provide suggestion for users on how to add their own logging without adding logging for any other external library, let alone all of them. So I basically don't see the origin of your reasoning that anybody suggested enabling logging for all libraries out there. If the users have additional libraries but haven't taken a step to enable them similar to the one for aexpect above, they will still not be logged.

> > The option I described above does exactly that: it allows the users to choose which libraries they want to include. Not allowing aexpect logging lines as lines outside of avocado is just as restrictive as demanding aexpect lines to be displayed when some of us might not want to see them for supposed high level logging output without that many details.
> 
> I have a feeling that changing their parent could be a little "intrusive" and not needed. With the Python logging system, maybe we don't need to change their parent to 'avocado.' or 'avocado.extlib'. We just need a 'logging.conf' (or an equivalent python code) that will redirect specific libraries to our handlers (files) of interest. Again, I'm talking about 'other libraries', aexpect might be a special case.

That would probably be even better indeed. Do you have a concrete example of logging configuration that could emulate the same behavior?

> 2\. We are teaching our users that everything that is external it is under `avocado.*`. By doing that we might assume some messages will be there just because currently it is. And that might be false, if at some point in the future any external library decided to change the log message. Users will miss that message or get a different message.

Indeed, coming to think of this right now I can't recall if we used the `avocado.extlib` in order to enable the logging within avocado or just for consistency reasons with respect to attaching the parent logger as we do in code.

> I guess each plugin is free to handle the logging the way they want. And I'm just trying to discuss here alternatives, in order to make things consistent with the design of the "new logging system" adopted inside Avocado. Otherwise we are going to have two different designs in two really close projects.

At least in that regard I don't think Avocado and Avocado VT are that similar, mostly because Avocado VT has many more dependencies in order to function properly. So in simplest terms, the way I imagine this is that each project will add to the avocado scope (and debug logs) logging for additional libraries that just it depends on and is aware of.

**Comment by beraldoleal** [comment]
> I see @beraldoleal but I don't quite get where we suggest adding other external libraries here. 

"In particular for Avocado VT, we have to do this **not just for aexpect but any other external library out there** ..."

Maybe I got something wrong....
 
> > > The option I described above does exactly that: it allows the users to choose which libraries they want to include. Not allowing aexpect logging lines as lines outside of avocado is just as restrictive as demanding aexpect lines to be displayed when some of us might not want to see them for supposed high level logging output without that many details.
> > 
> > 
> > I have a feeling that changing their parent could be a little "intrusive" and not needed. With the Python logging system, maybe we don't need to change their parent to 'avocado.' or 'avocado.extlib'. We just need a 'logging.conf' (or an equivalent python code) that will redirect specific libraries to our handlers (files) of interest. Again, I'm talking about 'other libraries', aexpect might be a special case.
> 
> That would probably be even better indeed. Do you have a concrete example of logging configuration that could emulate the same behavior?

This is not tested, and probably it is broken. But the way I imagine this is something like this:

```ini
[loggers]
keys=avocado,aexpect

[handlers]
keys=consoleHandler

[formatters]
keys=simpleFormatter

[logger_avocado]
level=DEBUG
handlers=consoleHandler

[logger_aexpect]
level=DEBUG
handlers=consoleHandler
qualname=aexpect
propagate=0

[handler_consoleHandler]
class=StreamHandler
level=DEBUG
formatter=simpleFormatter
args=(sys.stdout,)

[formatter_simpleFormatter]
format=%(asctime)s - %(name)s - %(levelname)s - %(message)s
```

Please, take it with a grain of salt: this is just a simple version with only one handler (which is not debug.log) and for future implementation. We need to think on a solution for the @luckyh problem first. But the idea is that users, tweaking config files, would be able to merge loggers using the default `logging.config` syntax.

**Comment by luckyh** [comment]
> What about using `avocado.aexpect.something` as the namespace for the aexpect modules? Since aexpect is part of Avocado project, I don't see a problem with that.
> 
> Would that fix the issue?

@beraldoleal yes, this is indeed a simple way to fix the issue. I just tried with one of the aexpect module (`aexpect.client`) and it worked.

```
[stdlog] 2022-03-15 22:20:32,166 avocado.virttest.virt_vm DEBUG| Found/Verified IP 192.168.1.6 for VM avocado-vt-vm1 NIC 0
[stdlog] 2022-03-15 22:20:34,404 avocado.aexpect.client DEBUG| Sending command: ls
[stdlog] 2022-03-15 22:20:34,517 avocado.aexpect.client DEBUG| Sending command: echo $?
[stdlog] 2022-03-15 22:20:34,780 avocado.virttest.qemu_vm DEBUG| Destroying VM avocado-vt-vm1 (PID 1732783)
```

But, please let me confirm that anyone in the discussion is happy with having this approach, @beraldoleal @pevogam I'd like to know your opinion, thanks!

**Comment by luckyh** [comment]
> Hi @luckyh can you please send how you are running your solution number 1. Because this is not the desire behavior how the `–store-logging-stream` should work, and I would like to investigate it. Thanks.

@richtja I ran the following command with the modified `boot` test.
```
avocado run --store-logging-stream=aexpect --vt-type=qemu -- io-github-autotest-qemu.boot
```
**tp-qemu**
```patch
diff --git a/generic/tests/boot.py b/generic/tests/boot.py
index 9b26637f..c01468ba 100644
--- a/generic/tests/boot.py
+++ b/generic/tests/boot.py
@@ -27,6 +27,7 @@ def run(test, params, env):
             session = vm.wait_for_serial_login(timeout=timeout)
         else:
             session = vm.wait_for_login(timeout=timeout)
+        session.cmd('ls')
         session.close()
 
     if params.get("rh_perf_envsetup_script"):
```
With the feature enabled we could see these lines from the debug.log.
```
[stderr] DEBUG:aexpect.client:Sending command: ls
[stderr]
[stderr] DEBUG:aexpect.client:Sending command: echo $?
[stderr]
```

**Comment by luckyh** [comment]
> As aexpect could be used in standalone applications (we have some usage like that) and doesn't have any explicit direct dependency on Avocado, I would recommend not mixing the namespaces any further. What we do on our side to make the aexpect messages available again (but we needed the namespace [fix above](https://github.com/avocado-framework/aexpect/pull/89) first) is
> 
> ```python
> log.getLogger("aexpect").parent = log.getLogger("avocado.extlib")
> ```
> 
> i.e. we bind the loggers dynamically. In particular for Avocado VT, we have to do this not just for aexpect but any other external library out there (we attach all of them to the avocado.extlib namespace above) and do this using a VT test runner customization hook from #3242. So I consider our solution an approach (3) to the two approaches above, and at least from what I can see it seems the minimal amount of changes.
> 
> @luckyh What do you think about this third approach? All the needed adaptations to make it work are already in place.

@pevogam indeed, it looks better, but my concern is that the third approach was still on the hard-coding way more or less. Also, which is the most important thing from my perspective, I think Avocado should provide a **unified** method (perhaps by introducing some APIs or configurations) for plugins/endpoint-apps to handle the logging stuff rather than let plugins/endpoint-apps implement that by themselves, and that's why I was looking for such and discovered the first approach (although it is not the desired one said by @richtja).

**Comment by luckyh** [comment]
> > > I have a feeling that changing their parent could be a little "intrusive" and not needed. With the Python logging system, maybe we don't need to change their parent to 'avocado.' or 'avocado.extlib'. We just need a 'logging.conf' (or an equivalent python code) that will redirect specific libraries to our handlers (files) of interest. Again, I'm talking about 'other libraries', aexpect might be a special case.
> > 
> > 
> > That would probably be even better indeed. Do you have a concrete example of logging configuration that could emulate the same behavior?
> 
> This is not tested, and probably it is broken. But the way I imagine this is something like this:
> 
> ```ini
> [loggers]
> keys=avocado,aexpect
> 
> [handlers]
> keys=consoleHandler
> 
> [formatters]
> keys=simpleFormatter
> 
> [logger_avocado]
> level=DEBUG
> handlers=consoleHandler
> 
> [logger_aexpect]
> level=DEBUG
> handlers=consoleHandler
> qualname=aexpect
> propagate=0
> 
> [handler_consoleHandler]
> class=StreamHandler
> level=DEBUG
> formatter=simpleFormatter
> args=(sys.stdout,)
> 
> [formatter_simpleFormatter]
> format=%(asctime)s - %(name)s - %(levelname)s - %(message)s
> ```

@beraldoleal this also looks good to me. Just one comment is that I'd suggest introducing Avocado's own logging configuration schema from a high-level point of view instead of directly using the one provided by Python, that's because I think it can:

1. hide the details users don't care about
2. protect the logging settings we don't want users to touch
3. easy to extend if Avocado has the plan to support other languages

**Comment by beraldoleal** [comment]
> @beraldoleal this also looks good to me. Just one comment is that I'd suggest introducing Avocado's own logging configuration schema from a high-level point of view instead of directly using the one provided by Python, that's because I think it can:
> 
>     hide the details users don't care about
>     protect the logging settings we don't want users to touch
>     easy to extend if Avocado has the plan to support other languages
> 

Here I guess we disagree a little bit. ;) Why reinvent the well if many python projects uses this scheme (Openstack, django, flask, ...) ?  Any external Python contributor or an user familiar with the python logging system it will be quickly onboard.  IMO this would reduce the learning curve. About your points:

>     hide the details users don't care about
 >     protect the logging settings we don't want users to touch

Ideally users don't need to touch on this file. We should ship a default and functional file for most of the cases. Only advanced users that are interested on other namespaces would be interested on changing this file. Why needing hard coded hacks to change behaviors if this could be done via a standard python way? 

Yes, we could still keep some command-line options to choose what to show and what to hide, like we have today. But IMO we should not recreated a entire config logic into code, because it will be harder for users to adapt, when working with particular cases.

>     easy to extend if Avocado has the plan to support other languages

I might be wrong here, but I don't see Avocado core components and plugins in a non-python language. Even that we support *tests* in other languages we will still need a runner on the Python side. And I see no point of having multiple logging systems for each language we support. The logging part it would be the runner's responsibility. (Again, I might be wrong, this is my biased opinion)

**Comment by luckyh** [comment]
> Here I guess we disagree a little bit. ;) Why reinvent the well if many python projects uses this scheme (Openstack, django, flask, ...) ? Any external Python contributor or an user familiar with the python logging system it will be quickly onboard. IMO this would reduce the learning curve.

Well, I would agree with that if we are talking about some developers/maintainers oriented logging implementation, for example  using a code based way just like [Django's](https://docs.djangoproject.com/en/4.0/topics/logging/#examples) could definitely reduce the learning curve for them. However, for users who especially may not familiar with Python's logging facility or even the Python language but just want to enable more log messages for their use case, I'm not sure if that will help them in a better way.

In fact, there are already some logging related options (e.g. [`job.output.loglevel`](https://avocado-framework.readthedocs.io/en/latest/config/index.html#job-output-loglevel)) presented in Avocado's configuration, and thus I suggested the way I mentioned in my previous comment, compared with **directly** using the Python's one which I think it is a low-level way and may not be friendly to end users.

Actually, I think I just suggested some wrapping work instead of reinventing the wheel, for example, the low-level code/configuration in Python's logging config scheme could be dynamically generated by respecting Avocado's configuration.

> Ideally users don't need to touch on this file. We should ship a default and functional file for most of the cases. Only advanced users that are interested on other namespaces would be interested on changing this file.
> 
>...
>
> Yes, we could still keep some command-line options to choose what to show and what to hide, like we have today. But IMO we should not recreated a entire config logic into code, because it will be harder for users to adapt, when working with particular cases.

Then, yes, and moreover, the low-level code/configuration should never be modified by end users (my second point), only developers/maintainers can update it (by patching the code), while everyone can tweak the relevant part of Avocado's configuration.

> I might be wrong here, but I don't see Avocado core components and plugins in a non-python language. Even that we support _tests_ in other languages we will still need a runner on the Python side. And I see no point of having multiple logging systems for each language we support. The logging part it would be the runner's responsibility. (Again, I might be wrong, this is my biased opinion)

Okay, that makes sense to me if there always be a runner written in Python serving for that, thanks.

**Comment by richtja** [comment]
> > Hi @luckyh can you please send how you are running your solution number 1. Because this is not the desire behavior how the `–store-logging-stream` should work, and I would like to investigate it. Thanks.
> 
> @richtja I ran the following command with the modified `boot` test.
> 
> ```
> avocado run --store-logging-stream=aexpect --vt-type=qemu -- io-github-autotest-qemu.boot
> ```
> 
> **tp-qemu**
> 
> ```diff
> diff --git a/generic/tests/boot.py b/generic/tests/boot.py
> index 9b26637f..c01468ba 100644
> --- a/generic/tests/boot.py
> +++ b/generic/tests/boot.py
> @@ -27,6 +27,7 @@ def run(test, params, env):
>              session = vm.wait_for_serial_login(timeout=timeout)
>          else:
>              session = vm.wait_for_login(timeout=timeout)
> +        session.cmd('ls')
>          session.close()
>  
>      if params.get("rh_perf_envsetup_script"):
> ```
> 
> With the feature enabled we could see these lines from the debug.log.
> 
> ```
> [stderr] DEBUG:aexpect.client:Sending command: ls
> [stderr]
> [stderr] DEBUG:aexpect.client:Sending command: echo $?
> [stderr]
> ```

Thanks, @luckyh, for this information. I found out that in avocado we have an issue with custom loggers and `–store-logging-stream`. I created issue [#5290](https://github.com/avocado-framework/avocado/issues/5290), and I am working on a fix. I will let you know when the fix will be ready.

**Comment by pevogam** [comment]
@luckyh Can you confirm if this resolves your issue? I am just curious if this was all the changes it needed.

**Comment by luckyh** [comment]
> @luckyh Can you confirm if this resolves your issue? I am just curious if this was all the changes it needed.

Yes, I have tested the fix and it worked as expected (https://github.com/avocado-framework/avocado/pull/5292#issuecomment-1080305777), so here I think we are choosing the first approach given in the [desciption](https://github.com/avocado-framework/avocado-vt/issues/3369#issue-1168195060).

One thing I'm not sure is if the logging configuration which was mentioned in previous comments is something going to be supported by Avocado, but since it is not confirmed and not only for aexpect let's file another issue to track it if necessary, thanks.

**Comment by richtja** [comment]
@pevogam and @luckyh  there is an [issue](https://github.com/avocado-framework/avocado/issues/5178) on avocado which tracks the  logging configuration.

**Comment by luckyh** [comment]
> @pevogam and @luckyh there is an [issue](https://github.com/avocado-framework/avocado/issues/5178) on avocado which tracks the logging configuration.

@richtja that's great, thanks!

**Comment by pevogam** [comment]
I see, looking at the PR fixing this it seems that it was meant to work only with additional `--store-logging-stream` arguments. Will bring some of my further concerns there.
